File: 2021/Day16/Day16.scala
import scala.io.Source
import scala.annotation.tailrec

@main def day16 =

    extension (s: String)
        def |(start: Int) = s.slice(start, s.length)
        def ||(start: Int, end: Int) = s.slice(start, end + 1)

    enum Packet:
        case Literal(version: Int, value: Long)
        case Operation(version: Int, option: Int, subs: List[Packet])

    import Packet._

    val hex2bin = Map(
      ('0', "0000"),
      ('1', "0001"),
      ('2', "0010"),
      ('3', "0011"),
      ('4', "0100"),
      ('5', "0101"),
      ('6', "0110"),
      ('7', "0111"),
      ('8', "1000"),
      ('9', "1001"),
      ('A', "1010"),
      ('B', "1011"),
      ('C', "1100"),
      ('D', "1101"),
      ('E', "1110"),
      ('F', "1111")
    )

    val endPattern = "^[0]+$".r

    def endString(s: String) =
        endPattern.pattern.matcher(s).matches
            || s.length == 0

    def version(s: String) = Integer.parseInt(s || (0, 2), 2)

    def packageType(s: String) = Integer.parseInt(s || (3, 5), 2)

    def operation(s: String) = s(6).toString.toInt

    def literalLast(s: String) =
        @tailrec
        def loop(s: String, i: Int): Int =
            if s(i) == '0' then i + 4
            else loop(s, i + 5)

        loop(s, 0)

    def totalVersion(pkg: Packet) =
        def loop(pkg: Packet, acc: Int): Int =
            pkg match
                case Literal(v, _) => acc + v
                case Operation(v, _, ps) =>
                    acc + v + ps.foldLeft(0)((a, p) => a + loop(p, 0))

        loop(pkg, 0)

    def eval(pkg: Packet): Long =
        pkg match
            case Literal(_, value) => value
            case Operation(_, t, l) =>
                t match
                    case 0 => l.foldLeft(0L)((acc, p) => acc + eval(p))
                    case 1 => l.foldLeft(1L)((acc, p) => acc * eval(p))
                    case 2 => l.map(eval).min
                    case 3 => l.map(eval).max
                    case 5 =>
                        l match
                            case s :: f :: Nil =>
                                if eval(f) > eval(s) then 1 else 0
                            case _ => sys.error("Invalid type 5 payload")
                    case 6 =>
                        l match
                            case s :: f :: Nil =>
                                if eval(f) < eval(s) then 1 else 0
                            case _ => sys.error("Invalid type 6 payload")
                    case 7 =>
                        l match
                            case s :: f :: Nil =>
                                if eval(f) == eval(s) then 1 else 0
                            case _ => sys.error("Invalid type 7 payload")

    def buildPackage(s: String) =
        def literal(s: String) =
            val (v, idx) = (version(s), literalLast(s | 6))

            val value =
                for i <- 0 to idx if i % 5 != 0
                yield s(i + 6)

            (
              6 + idx + 1,
              Literal(v, java.lang.Long.parseLong(value.mkString, 2))
            )

        def loop(
            s: String,
            packets: List[Packet],
            limit: Option[Int]
        ): (String, List[Packet]) =
            (endString(s), limit) match
                case (true, _) | (_, Some(0)) => (s, packets)
                case _ =>
                    val (v, t, o) = (version(s), packageType(s), operation(s))
                    val lim = limit.map(_ - 1)
                    (t, o) match
                        case (4, _) =>
                            val (next, lit) = literal(s)
                            loop(s | next, lit :: packets, lim)
                        case (_, 0) =>
                            val range = Integer.parseInt(s || (7, 21), 2)

                            val (_, inner) =
                                loop(s || (22, 22 + range - 1), Nil, None)
                            val (next, other) = loop(s | (22 + range), Nil, lim)
                            (next, other ++ (Operation(v, t, inner) :: packets))
                        case _ =>
                            val innerL = Integer.parseInt(s || (7, 17), 2)

                            val (n, inner) = loop(s | 18, Nil, Some(innerL))
                            val (next, other) = loop(n, Nil, lim)
                            (next, other ++ (Operation(v, t, inner) :: packets))

        loop(s, Nil, Some(1))

    val input =
        Source
            .fromFile("input.txt")
            .getLines
            .next
            .toCharArray
            .map(hex2bin.get(_).get)
            .mkString

    val packet = buildPackage(input)._2.head

    println(s"First part: ${totalVersion(packet)}")
    println(s"Second part: ${eval(packet)}")

File: 2021/Day16/Day16.fs
module Day16

open System
open System.IO
open System.Text.RegularExpressions

type Packet =
    | Literal of int * int64
    | Operation of int * int * Packet list

let hex2bin =
    Map [ ('0', "0000")
          ('1', "0001")
          ('2', "0010")
          ('3', "0011")
          ('4', "0100")
          ('5', "0101")
          ('6', "0110")
          ('7', "0111")
          ('8', "1000")
          ('9', "1001")
          ('A', "1010")
          ('B', "1011")
          ('C', "1100")
          ('D', "1101")
          ('E', "1110")
          ('F', "1111") ]

let (<.>) f x y = f y x

let endString s =
    Regex.IsMatch(s, "^[0]+$")
    || s |> String.length = 0

let version (s: string) = Convert.ToInt32(s.[0..2], 2)

let packageType (s: string) = Convert.ToInt32(s.[3..5], 2)

let operation (s: string) = s.[6] |> string |> int

let literalLast s =
    let rec loop (s: string) i =
        if s.[i] = '0' then
            i + 4
        else
            loop s (i + 5)

    loop s 0

let totalVersion package =
    let rec loop package acc =
        match package with
        | Literal (v, _) -> acc + v
        | Operation (v, _, ps) -> acc + v + (ps |> List.sumBy (fun p -> loop p 0))

    loop package 0

let rec eval package =
    match package with
    | Literal (_, value) -> value
    | Operation (_, t, l) ->
        match t with
        | 0 -> l |> List.sumBy (fun p -> eval p)
        | 1 -> l |> List.fold (fun acc p -> acc * eval p) 1L
        | 2 -> l |> List.map eval |> List.min
        | 3 -> l |> List.map eval |> List.max
        | 5 ->
            match l with
            | [ s; f ] -> if eval f > eval s then 1 else 0
            | _ -> failwithf "Invalid type 5 payload %A" package
        | 6 ->
            match l with
            | [ s; f ] -> if eval f < eval s then 1 else 0
            | _ -> failwithf "Invalid type 6 payload %A" package
        | 7 ->
            match l with
            | [ s; f ] -> if eval f = eval s then 1 else 0
            | _ -> failwithf "Invalid type 7 payload %A" package
        | _ -> failwithf "Unsupported payload %A" package

let buildPackage s =
    let literal (s: string) =
        let v, idx = version s, literalLast s.[6..]

        let value =
            [| for i in 0 .. idx do
                   if i % 5 <> 0 then yield s.[i + 6] |]
            |> String

        6 + idx + 1, Literal(v, Convert.ToInt64(value, 2))

    let rec loop s packets limit =

        match endString s, limit with
        | true, _
        | _, Some (0) -> s, packets
        | _, _ ->
            let v, t, o = version s, packageType s, operation s
            let limit = limit |> Option.map ((-) <.> 1)

            match t, o with
            | 4, _ ->
                let next, literal = literal s
                loop s.[next..] (literal :: packets) limit
            | _, 0 ->
                let range = Convert.ToInt32(s.[7..21], 2)

                let _, inner = loop s.[22..22 + range - 1] [] None
                let next, other = loop s.[22 + range..] [] limit
                next, other @ Operation(v, t, inner) :: packets
            | _, _ ->
                let innerL = Convert.ToInt32(s.[7..17], 2)

                let next, inner = loop s.[18..] [] (Some innerL)
                let next, other = loop next [] limit
                next, other @ Operation(v, t, inner) :: packets

    loop s [] (Some 1)

let input =
    File.ReadAllLines "input.txt"
    |> Array.head
    |> Seq.map (fun s -> hex2bin |> Map.find s)
    |> String.concat ""
    |> (buildPackage >> snd >> List.head)

totalVersion input |> printfn "First part: %d"
eval input |> printfn "Second part: %d"
